### PR TITLE
refactor: optimize calldatasize check

### PR DESCRIPTION
**Pull request by emc415**
### What I did
This wip commit addresses the first half of inefficient selector table checks (issue #3037)

### How I did it
I added a function which tests whether a contract’s function has a selector with trailing zeroes (wip: at the moment it catches all selectors which contain 0)
It then changes the IR created in def _runtime_ir() to only have a line check in IR for calldatasize >= 4 if there is a zero selector function in the contract.

### Commit message

Addresses the first half of inefficient selector table checks (issue #3037) by only running a check for calldatasize >= 4 if a zero selector function is present.

### Description for the changelog
only check for calldatasize >= 4 if zero selector function is present.

### Cute Animal Picture

which is the animal? :D

Put a link to a cute animal picture inside the parenthesis-->

### CODE REVIEW

**Comment by emc415** [comment]
I figured out why my zero selector function wasn't getting pulled (and why I hacked it by finding any zero instead of trailing zeroes): it's because str(f._metadata["signature"]) is not pulling the abi_signature, just the signature. So for example, "def BlockHashAskewLimitary():" is getting hashed instead of "BlockHashAskewLimitary". Currently working on finding the correct function in the codebase to convert it to an abi signature.

File: vyper/codegen/module.py [hunk 1]
```diff
@@ -2,7 +2,7 @@
 
 from typing import Any, Dict, List, Optional, Tuple
 
-from vyper import ast as vy_ast
+from vyper import ast as vy_ast, utils
```
Comment (charles-cooper) [inline vyper/codegen/module.py:5]: it's best practice (and the linter will complain otherwise) for each import to have its own statement. i would also just import the specific functions needed from vyper.utils.

File: vyper/codegen/module.py [hunk 2]
```diff
@@ -91,6 +91,11 @@ def _is_internal(func_ast):
 def _is_payable(func_ast):
     return func_ast._metadata["type"].mutability == StateMutability.PAYABLE
 
+def _is_zero_selector_func(f):
+    abi_sig = f._metadata["signature"].abi_signature_for_kwargs([])
+    selector = utils.abi_method_id(abi_sig)
+    if str(selector)[0] == '0':
```
Comment (charles-cooper) [inline vyper/codegen/module.py:97]: probably want `if selector == b"\x00"*4` here, unless we decide we want the check to be for trailing zeroes. also above could use a sanity check that `len(selector) == 4`

File: vyper/codegen/module.py [hunk 3]
```diff
@@ -100,6 +105,10 @@ def _runtime_ir(runtime_functions, all_sigs, global_ctx):
     internal_functions = [f for f in runtime_functions if _is_internal(f)]
 
     external_functions = [f for f in runtime_functions if not _is_internal(f)]
+
+    # all functions with a zero selector, e.g. 0x00000000: changed location and functions referenced
```
Comment (charles-cooper) [inline vyper/codegen/module.py:109]: since we aren't actually using these functions, we can have this be a condition like `has_zero_selector_functions = any(is_zero_selector_func(f) for f in external_functions)`.

File: vyper/codegen/module.py [hunk 4]
```diff
@@ -160,15 +169,25 @@ def _runtime_ir(runtime_functions, all_sigs, global_ctx):
...
-    runtime = [
-        "seq",
-        # check that calldatasize is at least 4, otherwise
-        # calldataload will load zeros (cf. yellow paper).
-        ["if", ["lt", "calldatasize", 4], ["goto", "fallback"]],
-        ["with", "_calldata_method_id", shr(224, ["calldataload", 0]), selector_section],
-        close_selector_section,
-        ["label", "fallback", ["var_list"], fallback_ir],
-    ]
+    if len(zero_selector_functions) != 0:
```
Comment (charles-cooper) [inline vyper/codegen/module.py:172]: it would be good to merge the common parts of these two branches.

**Review by charles-cooper — COMMENTED** [review]
thanks for taking this on, nice work so far -- left a few comments in the review

File: vyper/codegen/module.py (hunk 1 above)
Comment (emc415) [reply vyper/codegen/module.py:5]: thanks!

File: vyper/codegen/module.py (hunk 3 above)
Comment (emc415) [reply vyper/codegen/module.py:109]: awesome!

File: vyper/codegen/module.py (hunk 2 above)
Comment (emc415) [reply vyper/codegen/module.py:97]: the selector is an int, once its been run through abi_method_id; therefore len(selector) == 4 will fail for the zero selector. I can run hex(selector) to get it in the form suggested, but the zero selector will still equal '0x0', not '0x00000000'. Perhaps it would be good to catch all selectors with trailing zeroes, as well?

Comment (emc415) [reply vyper/codegen/module.py:98]: fixed with:
if selector.to_bytes(4, 'big') == b"\x00"*4:

File: vyper/codegen/module.py (hunk 4 above)
Comment (emc415) [reply vyper/codegen/module.py]: updated in most recent commit

File: vyper/codegen/module.py [hunk 5]
```diff
@@ -91,6 +92,11 @@ def _is_internal(func_ast):
 def _is_payable(func_ast):
     return func_ast._metadata["type"].mutability == StateMutability.PAYABLE
 
+def _is_zero_selector_func(f):
+    abi_sig = f._metadata["signature"].abi_signature_for_kwargs([])
+    selector = abi_method_id(abi_sig)
+    if selector.to_bytes(4, 'big') == b"\x00"*4:
```
Comment (charles-cooper) [inline vyper/codegen/module.py:98]: bit weird, why not `return selector.to_bytes(4, "big") == b"\x00"*4`

or, `return selector == 0`?

File: vyper/codegen/module.py [hunk 6]
```diff
@@ -100,6 +106,10 @@ def _runtime_ir(runtime_functions, all_sigs, global_ctx):
     internal_functions = [f for f in runtime_functions if _is_internal(f)]
 
     external_functions = [f for f in runtime_functions if not _is_internal(f)]
+
+    # returns True if any function's selector has trailing 0s
+    has_zero_selector_functions = any(_is_zero_selector_func(f) for f in external_functions)
```
Comment (charles-cooper) [inline vyper/codegen/module.py:111]: so i think we need to check actually all possible methods for each function. so something like `ContractFunction.method_ids`.

File: vyper/codegen/module.py [hunk 7]
```diff
@@ -162,14 +172,16 @@ def _runtime_ir(runtime_functions, all_sigs, global_ctx):
...
-        ["if", ["lt", "calldatasize", 4], ["goto", "fallback"]],
         ["with", "_calldata_method_id", shr(224, ["calldataload", 0]), selector_section],
         close_selector_section,
         ["label", "fallback", ["var_list"], fallback_ir],
     ]
 
+    if has_zero_selector_functions:
+        # check that calldatasize is at least 4, otherwise
+        # calldataload will load zeros (cf. yellow paper).
+        runtime.insert(1, ["if", ["lt", "calldatasize", 4], ["goto", "fallback"]])
```
Comment (charles-cooper) [inline vyper/codegen/module.py:183]: i think the flow of this will be clearer if it's more like
```
runtime = ["seq"]
if has_zero_selector_functions:
  runtime.append(["if", ...])
runtime.extend(
  ...
)
```

File: vyper/codegen/function_definitions/external_function.py [hunk 8]
```diff
@@ -86,10 +86,12 @@ def handler_for(calldata_kwargs, default_kwargs):
         # a sequence of statements to strictify kwargs into memory
         ret = ["seq"]
 
-        # ensure calldata is at least of minimum length
-        args_abi_t = calldata_args_t.abi_type
-        calldata_min_size = args_abi_t.min_size() + 4
-        ret.append(["assert", ["ge", "calldatasize", calldata_min_size]])
+        # run check if 0 selector; otherwise selector table checks equality
```
Comment (charles-cooper) [inline vyper/codegen/function_definitions/external_function.py:89]: interesting. will have to think about this more

File: vyper/codegen/module.py (hunk 7 above)
Comment (emc415) [reply vyper/codegen/module.py:183]: oh, cool. thanks!

File: vyper/codegen/module.py (hunk 6 above)
Comment (emc415) [reply vyper/codegen/module.py]: changed with my latest commit - let me know if that's what you were thinking

File: vyper/codegen/module.py (hunk 7 above)
Comment (emc415) [reply vyper/codegen/module.py]: changed with my latest commit

File: vyper/codegen/function_definitions/external_function.py (hunk 8 above)
Comment (emc415) [reply vyper/codegen/function_definitions/external_function.py:89]: any further thoughts on this?

**Comment by fubuloubu** [comment]
@emc415 just convert back from Draft when ready (it's on the right hand side)

File: vyper/codegen/module.py [hunk 9]
```diff
@@ -100,6 +99,10 @@ def _runtime_ir(runtime_functions, all_sigs, global_ctx):
     internal_functions = [f for f in runtime_functions if _is_internal(f)]
 
     external_functions = [f for f in runtime_functions if not _is_internal(f)]
+
+    # returns True if any function's selector has trailing 0s
```
Comment (charles-cooper) [inline vyper/codegen/module.py:103]: two things -- this isn't a function so it doesn't return anything, also the code is not checking for trailing 0s, it's checking for exact equality to 0
```suggestion
    # check if any selector is 0
```

File: vyper/codegen/function_definitions/external_function.py (hunk 8 above)
Comment (charles-cooper) [reply vyper/codegen/function_definitions/external_function.py:89]: i don't think this is right. this elides the check in all cases except when the selector is 0.

Comment (emc415) [reply vyper/codegen/function_definitions/external_function.py:89]: are you saying that it should elide the check in all cases except when the selector has trailing zeroes?

"however when there are no trailing 0s, it can be omitted because the selector table already checks exact equality of the selector."

Comment (charles-cooper) [reply vyper/codegen/function_definitions/external_function.py:89]: nono, other way around

File: vyper/codegen/module.py (hunk 9 above)
Comment (emc415) [reply vyper/codegen/module.py]: thanks! implemented

File: vyper/codegen/function_definitions/external_function.py (hunk 8 above)
Comment (emc415) [reply vyper/codegen/function_definitions/external_function.py]: decided to remove this change. assert calldatasize lt 4 is not redundant to this check; so even if selector is 0 we want to keep this check.

File: vyper/codegen/module.py [hunk 10]
```diff
@@ -100,6 +100,12 @@ def _runtime_ir(runtime_functions, all_sigs, global_ctx):
     internal_functions = [f for f in runtime_functions if _is_internal(f)]
 
     external_functions = [f for f in runtime_functions if not _is_internal(f)]
+
+    # check if any selector is 0
+    has_zero_selector_functions = any(
+        [0 in ContractFunction.from_FunctionDef(f).method_ids.values() for f in external_functions]
```
Comment (charles-cooper) [inline vyper/codegen/module.py:106]: you don't need to call `ContractFunction.from_FunctionDef`. it's already on the function object -- see other uses like `func_ast._metadata["type"]`

**Review by charles-cooper — COMMENTED** [review]
this is looking pretty good. will try to merge in later this or next week.

**Comment by charles-cooper** [comment]
looking like there's a test failure in ctor assembly

**Comment by charles-cooper** [comment]
The failing test case is a little weird; it depends on JUMP being in the opcodes output to demarcate the initcode from the runtime code. I would instead use the assembly output which is a little more structured.

File: vyper/codegen/module.py [hunk 11]
```diff
@@ -168,6 +168,9 @@ def _runtime_ir(runtime_functions, all_sigs, global_ctx):
 
     runtime = ["seq"]
 
+    # function with 0 selector may get treated as the fallback fn without
+    # this check for calldatasize < 4. For all other fns this check is redundant
+    # because a similar check for calldatasize is done in the selector table.
```
[the remark this answers is not in the document]
Comment (charles-cooper) [reply vyper/codegen/module.py:173]: actually this is clarifying something for me, which is that for consistency, we need the calldatasize check for selectors with trailing 0s in addition to just straight 0 selector. otherwise those will get treated as asserts, not fallbacks

[the remark this answers is not in the document]
Comment (charles-cooper) [reply vyper/codegen/module.py:173]: the other option is just to skip the initial calldatasize check entirely, always, and just rely on the assertion in the selector table?

[the remark this answers is not in the document]
Comment (emc415) [reply vyper/codegen/module.py:173]: Is there a practical difference between going to fallback, versus relying on the assert? If the assert fails it should result in a REVERT, which is what the fallback function goes to also, right?

[the remark this answers is not in the document]
Comment (charles-cooper) [reply vyper/codegen/module.py:173]: well -- that's the default fallback function, but the user can supply their own.

[the remark this answers is not in the document]
Comment (emc415) [reply vyper/codegen/module.py]: thanks! submitted a go at a more in-depth explanation

File: vyper/codegen/module.py [hunk 12]
```diff
@@ -160,15 +166,35 @@ def _runtime_ir(runtime_functions, all_sigs, global_ctx):
...
-        ["with", "_calldata_method_id", shr(224, ["calldataload", 0]), selector_section],
-        close_selector_section,
-        ["label", "fallback", ["var_list"], fallback_ir],
-    ]
+    runtime = ["seq"]
+
+    # if there is a function whose selector is 0, this gets triggered
+    # instead of the fallback fn when 0 calldataload is supplied,
+    # b/c calldataload loads 0s past the end of physical calldata (cf. yellow paper),
+    # causing the 0 selector fn to be treated as the fallback fn.
```
Comment (charles-cooper) [inline vyper/codegen/module.py:174]: these 4 lines are a bit hard to parse. what does "this" mean on the first line?

File: vyper/codegen/module.py [hunk 13]
```diff
@@ -160,15 +166,35 @@ def _runtime_ir(runtime_functions, all_sigs, global_ctx):
...
+    # since supplying 0 calldata is expected to trigger the fallback fn,
+    # we check that calldatasize >= 4, which distinguishes the 0 selector
+    # from the fallback function "selector"
+    # (equiv. to "all selectors not in the selector table").
+    # for all other fns, calldatasize is already validated at entry
+    # to the fn to be at least >= 4, so we can elide the check.
```
Comment (charles-cooper) [inline vyper/codegen/module.py:180]: "at least >=" is redundant

File: vyper/codegen/module.py [hunk 14]
```diff
@@ -160,15 +166,35 @@ def _runtime_ir(runtime_functions, all_sigs, global_ctx):
...
+
+    # note: this check is also present for fns with a 0 selector,
+    # but the check is an assertion, rather than jumping to fallback
+    # if the assertion fails.
+    # note: fns with trailing 0s in the selector are not checked here
+    # b/c not providing the trailing 0s in calldata is considered bad practice,
```
Comment (charles-cooper) [inline vyper/codegen/module.py:186]: not just bad practice -- invalid method id

[the remark this answers is not in the document]
Comment (charles-cooper) [reply vyper/codegen/module.py:186]: thinking about this, it may actually be clearer (and results in a 1-2 instruction improvement in bytecode) to just check that calldatasize is not 0.

[the remark this answers is not in the document]
Comment (charles-cooper) [reply vyper/codegen/module.py:186]: in other words, `0x` (empty) calldata will go to fallback, but `0x00` (one byte, `b"\x00"`) of calldata will revert.

File: vyper/codegen/module.py [hunk 15]
```diff
@@ -160,15 +166,35 @@ def _runtime_ir(runtime_functions, all_sigs, global_ctx):
...
+    # and the contract should revert.
+    if has_zero_selector_functions:
+        runtime.append(["if", ["lt", "calldatasize", 4], ["goto", "fallback"]])
```
[the remark this answers is not in the document]
Comment (emc415) [reply vyper/codegen/module.py:189]: That is a good idea. Here are the cases I've come up with:
When 0x is supplied:
  the contract will go to fallback
If 0x0 is supplied:
  This is not valid bytecode and should not get accepted on the client's side
When 0x00 is supplied, or anything less than 4 bytes:
   the contract will revert in the selector table check
    (I can add an assert that calldata_min_size >= 4 to double check that args_abi_t.min_size() is not negative)
When 0x00000000 is supplied:
   The check in selector table will pass and then the 0 selector function will get appropriately treated as the 0 selector fn

Can you think of any others? If not I think this is safe to implement

[the remark this answers is not in the document]
Comment (charles-cooper) [reply vyper/codegen/module.py:189]: actually, idk. it incurs a calldatasize check at the beginning of the selector function. maybe cleaner is to add the check in the 0 selector entry itself. this will also push all the invariants into the same place in the codegen, so we don't need to reason about the invariants in different places going forward.

[the remark this answers is not in the document]
Comment (charles-cooper) [reply vyper/codegen/module.py:189]: so a carveout in the selector table -- if the selector is 0, instead of generating the usual `if (eq _method_id <method id literal>)`, we have `if (and (eq _method_id <method id literal, aka 0>) (gt calldatasize 0))`

[the remark this answers is not in the document]
Comment (emc415) [reply vyper/codegen/module.py]: I implemented this in the latest commit -- is this what you were going for?

File: vyper/codegen/function_definitions/external_function.py [hunk 16]
```diff
@@ -123,7 +123,26 @@ def handler_for(calldata_kwargs, default_kwargs):
...
+        # from the case where nil calldata is supplied, b/c calldataload loads
+        # 0s past the end of physical calldata (cf. yellow paper).
+        # since supplying 0 calldata is expected to trigger the fallback fn,
+        # we check that calldatasize > 0, which distinguishes the 0 selector
+        # from the fallback function "selector"
+        # (equiv. to "all selectors not in the selector table").
+
+        # note: fns with trailing 0s in the selector are not checked here
+        # b/c not providing the trailing 0s in calldata is considered invalid
+        # for specifying the method id, and the contract should revert
```
Comment (charles-cooper) [inline vyper/codegen/function_definitions/external_function.py:136]: i'm not sure this comment is required anymore -- or if it is there, it should point out that calldatasize well-formedness check is added in the function body as part of ABI validation.

File: vyper/codegen/function_definitions/external_function.py [hunk 17]
```diff
@@ -123,7 +123,26 @@ def handler_for(calldata_kwargs, default_kwargs):
...
+
+        if method_id.value == 0:
+            ret = [
+                "if",
+                ["and", ["eq", "_calldata_method_id", method_id], ["gt", "calldatasize", 0]],
+                ret,
+            ]
+        else:
+            ret = ["if", ["eq", "_calldata_method_id", method_id], ret]
```
Comment (charles-cooper) [inline vyper/codegen/function_definitions/external_function.py:145]: will be clearer as follows:
```suggestion
        method_id_check =  ["eq", "_calldata_method_id", method_id]
        # add clarifying comments here <--
        if method_id.value == 0:
            method_id_check = ["and",  ["gt", "calldatasize", 0], method_id_check]
        ret = ["if", method_id_check, ret]
```

File: vyper/codegen/function_definitions/external_function.py (hunk 16 above)
Comment (charles-cooper) [reply vyper/codegen/function_definitions/external_function.py:136]: "trailing 0s" is a bit confusing since it refers to the method id as opposed to the calldata. you could say, "cases where not enough calldata is supplied besides calldatasize == 0 ...".

File: vyper/codegen/function_definitions/external_function.py (hunk 17 above)
Comment (emc415) [reply vyper/codegen/function_definitions/external_function.py]: awesome! made that change

File: vyper/codegen/function_definitions/external_function.py (hunk 16 above)
Comment (emc415) [reply vyper/codegen/function_definitions/external_function.py]: yeah, in the latest commit I edited this comment with your suggestions, but I am also thinking we can just do away with it (starting at note:)

**Comment by charles-cooper** [comment]
looks decent. @fubuloubu what do you think?

**Comment by charles-cooper** [comment]
also tagging @wadeAlexC in case you want to leave any feedback.

**Review by charles-cooper — APPROVED** [review]

**Review by fubuloubu — APPROVED** [review]
